// whatwg-url/lib/dist/URL.dev.js
"use strict";

var conversions = require("webidl-conversions");

var utils = require("./utils.js");

var Impl = require(".//URL-impl.js");

var impl = utils.implSymbol;

function URL(url) {
  if (!this || this[impl] || !(this instanceof URL)) {
    throw new TypeError("Failed to construct 'URL': Please use the 'new' operator, this DOM object constructor cannot be called as a function.");
  }

  if (arguments.length < 1) {
    throw new TypeError("Failed to construct 'URL': 1 argument required, but only " + arguments.length + " present.");
  }

  var args = [];

  for (var i = 0; i < arguments.length && i < 2; ++i) {
    args[i] = arguments[i];
  }

  args[0] = conversions["USVString"](args[0]);

  if (args[1] !== undefined) {
    args[1] = conversions["USVString"](args[1]);
  }

  module.exports.setup(this, args);
}

URL.prototype.toJSON = function toJSON() {
  if (!this || !module.exports.is(this)) {
    throw new TypeError("Illegal invocation");
  }

  var args = [];

  for (var i = 0; i < arguments.length && i < 0; ++i) {
    args[i] = arguments[i];
  }

  return this[impl].toJSON.apply(this[impl], args);
};

Object.defineProperty(URL.prototype, "href", {
  get: function get() {
    return this[impl].href;
  },
  set: function set(V) {
    V = conversions["USVString"](V);
    this[impl].href = V;
  },
  enumerable: true,
  configurable: true
});

URL.prototype.toString = function () {
  if (!this || !module.exports.is(this)) {
    throw new TypeError("Illegal invocation");
  }

  return this.href;
};

Object.defineProperty(URL.prototype, "origin", {
  get: function get() {
    return this[impl].origin;
  },
  enumerable: true,
  configurable: true
});
Object.defineProperty(URL.prototype, "protocol", {
  get: function get() {
    return this[impl].protocol;
  },
  set: function set(V) {
    V = conversions["USVString"](V);
    this[impl].protocol = V;
  },
  enumerable: true,
  configurable: true
});
Object.defineProperty(URL.prototype, "username", {
  get: function get() {
    return this[impl].username;
  },
  set: function set(V) {
    V = conversions["USVString"](V);
    this[impl].username = V;
  },
  enumerable: true,
  configurable: true
});
Object.defineProperty(URL.prototype, "password", {
  get: function get() {
    return this[impl].password;
  },
  set: function set(V) {
    V = conversions["USVString"](V);
    this[impl].password = V;
  },
  enumerable: true,
  configurable: true
});
Object.defineProperty(URL.prototype, "host", {
  get: function get() {
    return this[impl].host;
  },
  set: function set(V) {
    V = conversions["USVString"](V);
    this[impl].host = V;
  },
  enumerable: true,
  configurable: true
});
Object.defineProperty(URL.prototype, "hostname", {
  get: function get() {
    return this[impl].hostname;
  },
  set: function set(V) {
    V = conversions["USVString"](V);
    this[impl].hostname = V;
  },
  enumerable: true,
  configurable: true
});
Object.defineProperty(URL.prototype, "port", {
  get: function get() {
    return this[impl].port;
  },
  set: function set(V) {
    V = conversions["USVString"](V);
    this[impl].port = V;
  },
  enumerable: true,
  configurable: true
});
Object.defineProperty(URL.prototype, "pathname", {
  get: function get() {
    return this[impl].pathname;
  },
  set: function set(V) {
    V = conversions["USVString"](V);
    this[impl].pathname = V;
  },
  enumerable: true,
  configurable: true
});
Object.defineProperty(URL.prototype, "search", {
  get: function get() {
    return this[impl].search;
  },
  set: function set(V) {
    V = conversions["USVString"](V);
    this[impl].search = V;
  },
  enumerable: true,
  configurable: true
});
Object.defineProperty(URL.prototype, "hash", {
  get: function get() {
    return this[impl].hash;
  },
  set: function set(V) {
    V = conversions["USVString"](V);
    this[impl].hash = V;
  },
  enumerable: true,
  configurable: true
});
module.exports = {
  is: function is(obj) {
    return !!obj && obj[impl] instanceof Impl.implementation;
  },
  create: function create(constructorArgs, privateData) {
    var obj = Object.create(URL.prototype);
    this.setup(obj, constructorArgs, privateData);
    return obj;
  },
  setup: function setup(obj, constructorArgs, privateData) {
    if (!privateData) privateData = {};
    privateData.wrapper = obj;
    obj[impl] = new Impl.implementation(constructorArgs, privateData);
    obj[impl][utils.wrapperSymbol] = obj;
  },
  "interface": URL,
  expose: {
    Window: {
      URL: URL
    },
    Worker: {
      URL: URL
    }
  }
};
